### PR TITLE
core: tee: move tee entry_std.c to core/tee

### PR DESCRIPTION
**Pull request by maroueneboubakri**
Code in `entry_std.c`  is not architecture-specific and could be used by a new architecture port as it is. To avoid code duplication move the source file to `core/tee ` folder.

### CODE REVIEW

File: core/tee/sub.mk [hunk 1]
```diff
@@ -17,6 +17,8 @@ CFG_CRYPTO_PBKDF2 ?= y
 
 endif
 
+srcs-y += entry_std.c
+
```
Comment (jforissier) [inline core/tee/sub.mk:25]: This blank line is not really helpful

**Review by jforissier — COMMENTED** [review]
Hi @maroueneboubakri,
Thanks for the patches, LGTM with one very minor thing below.
`Acked-by: Jerome Forissier <jerome@forissier.org>`

**Comment by jenswi-linaro** [comment]
Does it really make sense to move entry_fast.[ch]? That one is quite Arm centric with all the SMC ids.

**Comment by maroueneboubakri** [comment]
> Does it really make sense to move entry_fast.[ch]? That one is quite Arm centric with all the SMC ids.

The port can still 'emulate' fast calls. At least to avoid making many changes on the OP-TEE linux kernel driver.

**Comment by jenswi-linaro** [comment]
To me it looks like you're better off without those. I imagine that most of them will be quite useless. Regardless you'll need a different probe function for the driver for Risc-V.

At some point you're really going to want to get rid of those SMCs, when you're trying to upstream the driver for instance. What happens then? Will we move entry_fast.[ch] back to arch/arm/tee?

**Comment by maroueneboubakri** [comment]
> To me it looks like you're better off without those. I imagine that most of them will be quite useless. Regardless you'll need a different probe function for the driver for Risc-V.
> 
> At some point you're really going to want to get rid of those SMCs, when you're trying to upstream the driver for instance. What happens then? Will we move entry_fast.[ch] back to arch/arm/tee?

These are needed at least while probing OP-TEE during Linux kernel boot. 
[](https://github.com/torvalds/linux/blob/v5.10/drivers/tee/optee/core.c#L615)

Basically this prevents being much intrusive when modifying the kernel driver. We keep it as it is, we only change the SMC invoke function. basically instead of `smc` instruction we invoke RISC-V environment call instruction, `ecall`.  
At SM level we identify the call type as per ATF and jump to the appropriate handler in OP-TEE. In OP-TEE the same vector table is provided to the SM as well.

**Comment by jenswi-linaro** [comment]
I can see how it's convenient for your while doing the initial port, those are the typical tricks you might do in out of tree changes in order to avoid conflicts when rebasing. However, I'm not sure this is what we'll use on the longer term. Surely you'd like to avoid depending on the Arm SMC Calling Convention on a Risc-V platform.

entry_fast.c depends on:
- struct thread_smc_args from core/arch/arm/include/kernel/thread.h
- core/arch/arm/include/sm/optee_smc.h
- core/arch/arm/include/kernel/tee_l2cc_mutex.h
- core/arch/arm/include/kernel/misc.h
- core/arch/arm/include/kernel/boot.h

How are you going to deal with those?

**Comment by maroueneboubakri** [comment]
The goal for the initial port was to be SMCCC compatible and stay as close as possible to what we're using today. [Referring to this comment](https://github.com/OP-TEE/optee_os/issues/4231#issuecomment-735649743).

`struct thread_smc_args` is kept as it is no change required.
`core/arch/arm/include/sm/optee_smc.h` kept as it is
ARM related flags are forced to `no `such as `CFG_PL310`. Related functions such returns not implemented or unknown. 

This has been done with [another port](https://github.com/op-tee-ia/optee_os/blob/master/core/arch/x86_64/tee/entry_fast.c)

**Comment by jenswi-linaro** [comment]
> The goal for the initial port was to be SMCCC compatible and stay as close as possible to what we're using today. [Referring to this comment](https://github.com/OP-TEE/optee_os/issues/4231#issuecomment-735649743).

OK, now I understand. In that case I'd prefer if entry_fast.[ch] are kept where they are. The few trivial functions you need can be copied into `core/arch/riscv`. In case you need many defines from optee_smc.h (I guess you'll need) can probably have optee_smc.h in `core/include` instead.

> 
> `struct thread_smc_args` is kept as it is no change required.
> `core/arch/arm/include/sm/optee_smc.h` kept as it is
> ARM related flags are forced to `no `such as `CFG_PL310`. Related functions such returns not implemented or unknown.
> 
> This has been done with [another port](https://github.com/op-tee-ia/optee_os/blob/master/core/arch/x86_64/tee/entry_fast.c)

Thanks for the link, hadn't seen this before.

**Comment by jenswi-linaro** [comment]
`Reviewed-by: Jens Wiklander <jens.wiklander@linaro.org>`

**Comment by maroueneboubakri** [comment]
> > The goal for the initial port was to be SMCCC compatible and stay as close as possible to what we're using today. [Referring to this comment](https://github.com/OP-TEE/optee_os/issues/4231#issuecomment-735649743).
> 
> OK, now I understand. In that case I'd prefer if entry_fast.[ch] are kept where they are. The few trivial functions you need can be copied into `core/arch/riscv`. In case you need many defines from optee_smc.h (I guess you'll need) can probably have optee_smc.h in `core/include` instead.
> 
> > `struct thread_smc_args` is kept as it is no change required.
> > `core/arch/arm/include/sm/optee_smc.h` kept as it is
> > ARM related flags are forced to `no `such as `CFG_PL310`. Related functions such returns not implemented or unknown.
> > This has been done with [another port](https://github.com/op-tee-ia/optee_os/blob/master/core/arch/x86_64/tee/entry_fast.c)
> 
> Thanks for the link, hadn't seen this before.

Good plan, agree.

**Pull request merged**